File: src/pcapack/src/rand_svd.f90
! this source code form is subject to the terms of the mozilla public
! license, v. 2.0. if a copy of the mpl was not distributed with this
! file, you can obtain one at http://mozilla.org/mpl/2.0/.


! copyright 2013, schmidt


! purpose
! =======
! 
! 
!
!
! notes
! =====
!
! the randomized svd method from the paper halko, martinsson, and tropp
! 2011, "finding structure with randomness: probabilistic algorithms 
! for constructing approximate matrix decompositions", siam review 
! 53, 217-288
!
! we implement the method from the paper in 2 ways.  the first is 
! "as is", using a random normal matrix generation for the test
! matrix omega.  additionally, we use the fastmap algorithm to 
! generate the matrix omega.
!
! algorithm is o(m*n*k); see the paper for details.
!
!
! arguments
! =========
!
!  m, n     (input) integer
!           the number of rows and columns of the input matrix a.
!
!  a        (input) double precision array
!           input data matrix of size mxn.
!
!  k        (input) integer
!           the desired number of singular vectors.
!
!  u        (output) double precision array
!           mxk
!
!  q        (input) integer
!           
!


! s = min(m, n)
! u = mxk
! vt = 

! omega = nxl, we can take l=2*k for sizes george used
! maybe try using fastmap to create omega


subroutine drandsvd(method, jobu, jobvt, m, n, a, k, q, s, u, vt, info)
  implicit none
  ! in/out
  character*1         method, jobu, jobvt
  integer :: m, n, k, q, info
  double precision :: a(*), s(*), u(*), vt(*)
  ! local
  integer :: i, lwork, rc_min, rc_max
  integer :: allocerr
  double precision :: tmp
  double precision, allocatable, dimension(:) ::  omega, work, y, atmp
  ! parameters
  double precision :: zero, one
  parameter ( zero = 0.0d0, one = 1.0d0 )
  ! subroutines
  intrinsic           min, max
  external            drandsvd_omega_init, drandsvd_stage_a, drandsvd_stage_b
  
  
  info = 0
  allocerr = 0
  
  
  !!! quick return if possible
  if (m < 1 .or. n < 1) then
    return
  end if
  
  if (method /= 'r' .or. method /= 'f') then
    info = -1
    return
  end if
  
  if (jobu /= 'v' .or. jobu /= 'n') then
    info = -2
    return
  end if
  
  if (jobvt /= 'v' .or. jobvt /= 'n') then
    info = -3
    return
  end if
  
  if (k.gt.n .or. k < 1) then
    info = -7
    return
  end if
  
  if (q < 1) then
    info = -9
    return
  end if
  
  rc_min = min(m, n)
  rc_max = max(m, n)
  
  
  !!! allocate and initializeomega=nx2k
  allocate(omega(n*2*k), stat=allocerr)
  call drandsvd_omega_init(method, n, k, omega, info)
  
  
  !!! allocate workspace
!      call dgeqrf(m, n, y, m, tau, tmp, -1, info)
!      lwork = int(tmp)
  ! fixme
  lwork = 10000
  allocate(work(lwork), stat=allocerr)
  
  
  !!! allocate 
  
  allocate(atmp(m*n), stat=allocerr)
  
  ! allocate y=mx2k
  allocate(y(m*2*k), stat=allocerr)
  
  
  !!! stage a
  call drandsvd_stage_a(m, n, k, q, a, y, u, omega, work, lwork, info)
  if (info /= 0) goto 1
  
  
  !!! stage b
  call drandsvd_stage_b(jobu, jobvt, m, n, k, a, y, s, u, vt, work, lwork, info)
  if (info /= 0) goto 1
  
  
  
  !!! free and return
1    continue
  
  write (*,*) info
  
  deallocate(omega)
  deallocate(work)
  deallocate(atmp)
  deallocate(y)
  
  
  return
end




! ------------------------------------------------------
! internal subroutines
! ------------------------------------------------------


! initialize omega, such as via random normal generation.
subroutine drandsvd_omega_init(method, n, k, omega, info)
  implicit none
  ! in/out
  character*1         method
  integer :: n, k, info
  double precision :: omega(*)
  ! subroutines
  external            rnormn
  
  
  ! random (standard) normal initialization
  if (method == 'r') then
    call rnormn(n*2*k, 0.0d0, 1.0d0, omega)
  ! fastmap initialization
  else if (method == 'f') then
    ! fixme
  ! bad argument 'method'
  else
    info = -1
    return
  end if
  
  
  return
end




! stage a from the paper, with the orthonormalization step
subroutine drandsvd_stage_a(m, n, k, q, a, y, u, omega, work, lwork, info)
  implicit none
  ! in/out
  integer :: m, n, k, q, lwork, info
  double precision :: a(*), y(*), u(*), omega(*), work(*)
  ! local
  integer :: j
  integer :: allocerr
  double precision, allocatable :: tau(:)
  ! parameters
  double precision :: zero, one
  parameter ( zero = 0.0d0, one = 1.0d0 )
  ! subroutines
  external            dgemm, dgesvd, drandsvd_subspaceiter
  
  
  allocate(tau(k), stat=allocerr)
  
  ! y = a * omega
  call dgemm('n', 'n', m, 2*k, n, one, a, m, omega, n, zero, y, m)
  
  ! q = the q part of the qr of y
  call dgeqrf(m, n, y, m, tau, work, lwork, info)
  if (info /= 0) goto 1
  call dorgqr(m, n, k, y, m, tau, work, lwork, info)
  if (info /= 0) goto 1
  
  
  !!! main loop
  
  ! i = 1
  call drandsvd_subspaceiter(m, n, k, q, a, u, y, tau, work, lwork, info)
  if (info /= 0) goto 1
  
  
  ! general
  do j = 2, q
    call drandsvd_subspaceiter(m, n, k, q, a, u, y, tau, work, lwork, info)
    if (info /= 0) goto 1
    
  end do
  
  
  ! free and return
1    continue
  
  deallocate(tau)
  
  return
end



subroutine drandsvd_subspaceiter(m, n, k, q, a, u, y, tau, work, lwork, info)
  implicit none
  ! in/out
  integer :: m, n, k, q, lwork, info
  double precision :: a(*), u(*), y(*), tau(*), work(*)
  ! local
  !
  ! parameters
  double precision :: zero, one
  parameter ( zero = 0.0d0, one = 1.0d0 )
  ! subroutines
  external            dgemm, dgeqrf, dorgqr
  
  
  ! twy_j = a^t * q_j-1
  call dgemm('t', 'n', n, 2*k, n, one, a, m, y, n, zero, y, m)
  
  ! qr of twy_j
  call dgeqrf(m, n, y, m, tau, work, lwork, info)
  if (info /= 0) return
  call dorgqr(m, n, k, y, m, tau, work, lwork, info)
  if (info /= 0) return
  
  ! y = a * q
  call dgemm('n', 'n', m, 2*k, n, one, a, m, y, n, zero, y, m)
  
  ! q = the q part of the qr of y
  call dgeqrf(m, n, y, m, tau, work, lwork, info)
  if (info /= 0) return
  call dorgqr(m, n, k, y, m, tau, work, lwork, info)
  if (info /= 0) return
  
  
  return
end




subroutine drandsvd_stage_b(jobu, jobvt, m, n, k, a, y, s, u, vt, work, lwork, info)
  implicit none
  ! in/out
  character*1         jobu, jobvt
  integer :: m, n, k, info
  double precision :: a(*), y(*), s(*), u(*), vt(*), work(*)
  ! local
  integer :: lwork, qr, qc, br, bc, ur, uc, rc_min
  integer :: allocerr
  double precision, allocatable :: b(:)
  ! parameters
  double precision :: zero, one
  parameter ( zero = 0.0d0, one = 1.0d0 )
  ! subroutines
  external            dgemm, dgesvd
  
  
  rc_min = min(m, n)
  
  br = min(rc_min, 2*k)
  bc = n
  
  qr = m
  qc = bc
  
  ur = m
  uc = min(n, 2*k)
  
  allocate(b(qc * n), stat=allocerr)
  if (allocerr /= 0) goto 1
  
  
  ! b = q^t * a
  call dgemm('t', 'n', qc, n, qr, one, y, qr, a, m, zero, b, qc)
  
  
  ! twu = the u part of the svd of b
  if (jobu == 'v') then
    call dgesvd('o', jobvt, qc, n, b, qc, s, 1.0d0, 1, vt, n, work, lwork, info)
    
    ! u = q * twu
    call dgemm('n', 'n', qr, qc, qc, one, y, qr, b, qc, zero, u, m)
    
    
    ! u = u[, 1:k]
    
    
    if (info /= 0) goto 1
  else
    call dgesvd('n', jobvt, qc, n, b, qc, s, 1.0d0, 1, vt, n,  work, lwork, info)
    
    
    ! vt = vt[1:k, ]
    
    
    if (info /= 0) goto 1
  end if
  
  
  
1    continue
  deallocate(b)
  
  
  return
end

File: src/pcapack/src/randsvd.c
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

// Copyright 2015, Schmidt


#include "pcapack.h"
#include "lapack.h"
#include "misc.h"


// Input x is overwritten by its Q
static inline int get_qr(const int m, const int n, double *x, double *tau, double *work, const int lwork)
{
  int info = 0;
  const int rank = MIN(m, n);
  
  // q = q part or the qr of y
  dgeqrf_(&m, &n, x, &m, tau, work, &lwork, &info);
  if (info != 0) return info;
  dorgqr_(&m, &n, &rank, x, &m, tau, work, &lwork, &info);
  
  return info;
}



//    B <- t(Q) %*% x
//    
//    if (!compute.u)
//      nu <- 0
//    else
//      nu <- min(nrow(B), ncol(B))
//    
//    if (!compute.vt)
//      nv <- 0
//    else
//      nv <- min(nrow(B), ncol(B))
//    
//    svd.B <- La.svd(x=B, nu=nu, nv=nv)
//    call svd(nu, nv, m, n, x, s, u, vt, info)
//    
//  end subroutine



//  k        (input) integer
//           the desired number of singular vectors.
//  u        (output) double precision array
//           mxk

// s = min(m, n)
// u = mxk
// vt = 

// omega = nxl, we can take l=2*k for sizes george used
// maybe try using fastmap to create omega



#define PCAPACK_RANDSVD_RANDINIT    1
#define PCAPACK_RANDSVD_FASTMAPINIT 2
#define PCAPACK_RANDSVD_BADMETHOD   -3




/**
 * @file
 * @brief Randsvd
 *
 * @details
 * The randomized svd method from the paper Halko, Martinsson, and Tropp
 * 2011, "Finding Structure with Randomness: Probabilistic Algorithms 
 * for Constructing Approximate Matrix Decompositions", SIAM Review 
 * 53, 217-288.
 * 
 * We implement the method from the paper in 2 ways.  The first is 
 * "as is", using a random normal matrix generation for the test
 * matrix omega.  Additionally, we use the fastmap algorithm to 
 * generate the matrix omega.
 * 
 * Complexity is O(m*n*k); see the paper for details.
 *
 * @param m,n
 * Inputs.  Problem size (dims of x)
 * @param x
 * 
 *
 * @return
 * The return value indicates that status of the function.  Non-zero values
 * are errors.
*/
int pcapack_randsvd(const bool retu, const bool retvt, const int k, const int niter, const int method, const int m, const int n, double *restrict x, double *restrict s, double *restrict u, double *restrict vt)
{
  int info = 0;
  int i, j, lwork, rc_min, rc_max;
  int nu, nv;
  double tmp;
  double *omega, *work, *Y, *tau, *B;
  const int twok = 2*k;
  
  
  if (m < 1 || n < 1) return 0;
  if (k > n || k < 1) return -7;
  if (niter < 1) return -9;
  
  if (method != PCAPACK_RANDSVD_RANDINIT && method != PCAPACK_RANDSVD_FASTMAPINIT)
    return PCAPACK_RANDSVD_BADMETHOD;
  
  rc_min = MIN(m, n);
  rc_max = MAX(m, n);
  
  // allocate and initializeomega=nx2k
  omega = malloc(n * twok * sizeof(*omega));
  
  if (method == PCAPACK_RANDSVD_RANDINIT)
  {
    // TODO use custom RNG
    for (i=0; i<n*twok; i++)
      omega[i] = rand()%2;
  }
  else if (method == PCAPACK_RANDSVD_FASTMAPINIT)
  {
    // FIXME
  }
  
  
  // Allocate workspace
  dgeqrf_(&m, &twok, Y, &m, tau, &tmp, &(int){-1}, &info);
  lwork = (int) tmp;
  
  work = malloc(lwork * sizeof(*work));
  Y = malloc(m * twok * sizeof(*Y));
  B = malloc(twok * n * sizeof(*B));
  
  // ----------------------------------------------------
  // Stage A from the paper with orthonormalization step
  // ----------------------------------------------------
  
  // Y = A*omega
  matmult(false, false, m, n, x, n, twok, omega, Y);
  
  // Y is overwritten by its QR's Q
  info = get_qr(m, n, Y, tau, work, lwork);
  if (info != 0) goto cleanup;
  
  
  // subspace iteration
  for (j=1; j<=niter; j++)
  {
    // twY_j = x^t * Q_j-1
    matmult(false, false, m, n, x, m, twok, Y, Y); // FIXME this is wrong...
    
    // qr of twY_j
    info = get_qr(m, n, Y, tau, work, lwork);
    
    // Y = x * Q
    matmult(false, false, m, n, x, m, twok, Y, Y); // FIXME this too...
    
    // Q = the Q part of the qr of Y
    info = get_qr(m, n, Y, tau, work, lwork);
    
    if (info != 0) goto cleanup;
  }
  
  
  // ----------------------------------------------------
  // Stage B from the paper
  // ----------------------------------------------------
  
  
  // B is 2KxN
  // B = t(Y) * x
  matmult(true, false, m, twok, Y, m, n, x, B);
  
  if (!retu) 
    nu = 0;
  else
    nu = MIN(twok, n);
  
  if (!retvt)
    nv = 0;
  else
    nv = MIN(twok, n);
  
  info = pcapack_svd(true, nu, nv, twok, n, B, s, u, vt);
  if (info != 0) goto cleanup;
  
  
  /*
  if (retu)
  {
    u <- svd.B$u
    u <- Q %*% U
    
    d <- svd.B$d
    
    d <- d[1L:k]
    u <- u[, 1L:k]
  }
  
  if (retvt)
  {
    vt <- svd.B$vt[1L:k, ]
  }
  */
  
  
  // ----------------------------------------------------
  // Cleanup and exit
  // ----------------------------------------------------
  
  cleanup:
    free(tau);
    free(Y);
    free(work);
    free(omega);
    free(B);
  
  return info;
}

